shared/src/commonTest/kotlin/co/touchlab/kampkit/BreedViewModelTest.kt

```diff
@@ -0,0 +1,258 @@
+            assertEquals(
+                BreedViewState(error = "Unable to download breed list"),
+                awaitItemPrecededBy(BreedViewState(isLoading = true), BreedViewState(isEmpty = true))
+            )
+        }
+    }
+
+    @Test
+    fun `Ignore API error with cache`() = runBlocking {
+        dbHelper.insertBreeds(breedNames)
+        settings.putLong(BreedRepository.DB_TIMESTAMP_KEY, (clock.currentInstant - 2.hours).toEpochMilliseconds())
+        ktorApi.throwOnCall(RuntimeException("Test error"))
+
+        viewModel.breedState.test {
+            assertEquals(
+                breedViewStateSuccessNoFavorite,
+                awaitItemPrecededBy(BreedViewState(isLoading = true))
+            )
+            expectNoEvents()
+
+            ktorApi.prepareResult(ktorApi.successResult())
+            viewModel.refreshBreeds().join()
+
+            assertEquals(
+                breedViewStateSuccessNoFavorite,
+                awaitItemPrecededBy(breedViewStateSuccessNoFavorite.copy(isLoading = true))
+            )
+        }
+    }
+
+    @Test
+    fun `Ignore API error on refresh with cache`() = runBlocking {
+        ktorApi.prepareResult(ktorApi.successResult())
+
+        viewModel.breedState.test {
+            assertEquals(
+                breedViewStateSuccessNoFavorite,
+                awaitItemPrecededBy(BreedViewState(isLoading = true), BreedViewState(isEmpty = true))
+            )
+            expectNoEvents()
+
+            ktorApi.throwOnCall(RuntimeException("Test error"))
+            viewModel.refreshBreeds().join()
+
+            assertEquals(
+                breedViewStateSuccessNoFavorite,
+                awaitItemPrecededBy(breedViewStateSuccessNoFavorite.copy(isLoading = true))
+            )
+        }
+    }
+
+    @Test
+    fun `Show API error on refresh without cache`() = runBlocking {
+        settings.putLong(BreedRepository.DB_TIMESTAMP_KEY, clock.currentInstant.toEpochMilliseconds())
+        ktorApi.throwOnCall(RuntimeException("Test error"))
+
+        viewModel.breedState.test {
+            assertEquals(BreedViewState(isEmpty = true), awaitItemPrecededBy(BreedViewState(isLoading = true)))
+            expectNoEvents()
+
+            viewModel.refreshBreeds().join()
+            assertEquals(
+                BreedViewState(error = "Unable to refresh breed list"),
+                awaitItemPrecededBy(BreedViewState(isEmpty = true, isLoading = true))
+            )
+        }
+    }
+}
+
+// There's a race condition where intermediate states can get missed if the next state comes too fast.
+// This function addresses that by awaiting an item that may or may not be preceded by the specified other items
+private suspend fun FlowTurbine<BreedViewState>.awaitItemPrecededBy(vararg items: BreedViewState): BreedViewState {
+    var nextItem = awaitItem()
+    for (item in items) {
+        if (item == nextItem) {
+            nextItem = awaitItem()
+        }
+    }
+    return nextItem
+}
```

Comment:
Is this too hacky?

Reply:
I'm not 100% sure what it does, but I think you're OK.